### ISSUE TITLE
IDE freezes for ~10sec when creating solution folder in Main.sln

### ISSUE THREAD

**Issue by DavidKarlas** [mono/monodevelop#6699]
Based on profiling, it seems like problem is in:
https://github.com/mono/monodevelop/blob/04d7368f/main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs#L283
Repro for me is pretty simple, try adding few new solution folders inside Main.sln, and IDE will freezes for few seconds.

> VS bug [#755010](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/755010)

**Comment by DavidKarlas** [comment]
Relevant stacktrace:
```
11080 MonoDevelop.Ide.RootWorkspace/<>c__DisplayClass103_0:<NotifyItemAddedToSolution>b__0 (object,System.EventArgs) [{0x7fd352352248} + 0x3c] [/Users/vsts/agent/2.142.1/work/1/s/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/RootWorkspace.cs :: 1164u] (0x14db29230 0x14db29276) [0x7fd35ef0e520 - VisualStudio.exe]
! 11080 (wrapper delegate-invoke) <Module>:invoke_void_object_SolutionItemChangeEventArgs (object,MonoDevelop.Projects.SolutionItemChangeEventArgs) [{0x7fd367554240} + 0xc2]  (0x12a8bf360 0x12a8bf4a3) [0x7fd35ef0e520 - VisualStudio.exe]
!   11076 MonoDevelop.VersionControl.VersionControlService:OnEntryAdded (object,MonoDevelop.Projects.SolutionItemEventArgs) [{0x7fd364687940} + 0x28b] [/Users/vsts/agent/2.142.1/work/1/s/monodevelop/main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlService.cs :: 608u] (0x14db29670 0x14db29b3b) [0x7fd35ef0e520 - VisualStudio.exe]
!   : 11076 MonoDevelop.VersionControl.Repository:GetDirectoryVersionInfo (MonoDevelop.Core.FilePath,bool,bool) [{0x7fd364686360} + 0x6e] [/Users/vsts/agent/2.142.1/work/1/s/monodevelop/main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs :: 285u] (0x12cf8dac0 0x12cf8dd98) [0x7fd35ef0e520 - VisualStudio.exe]
!   :   11076 System.Threading.WaitHandle:WaitOne () [{0x7fd3601ed9b8} + 0x24] [/Users/builder/jenkins/workspace/build-package-osx-mono/master@2/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/waithandle.cs :: 222u] (0x1140c2150 0x1140c217e) [0x7fd35ef0e520 - VisualStudio.exe]
!   :     11076 System.Threading.WaitHandle:WaitOne (int,bool) [{0x7fd3601ed968} + 0x3b] [/Users/builder/jenkins/workspace/build-package-osx-mono/master@2/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/waithandle.cs :: 206u] (0x10b5f3b30 0x10b5f3bd3) [0x7fd35ef0e520 - VisualStudio.exe]
!   :       11076 System.Threading.WaitHandle:WaitOne (long,bool) [{0x7fd3601ede20} + 0x3b] [/Users/builder/jenkins/workspace/build-package-osx-mono/master@2/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/waithandle.cs :: 239u] (0x10b5f3bf0 0x10b5f3c35) [0x7fd35ef0e520 - VisualStudio.exe]
!   :         11076 System.Threading.WaitHandle:InternalWaitOne (System.Runtime.InteropServices.SafeHandle,long,bool,bool) [{0x7fd3601ede68} + 0x4b] [/Users/builder/jenkins/workspace/build-package-osx-mono/master@2/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/waithandle.cs :: 250u] (0x10b5f3c50 0x10b5f3d1b) [0x7fd35ef0e520 - VisualStudio.exe]
!   :           11076 System.Threading.WaitHandle:WaitOneNative (System.Runtime.InteropServices.SafeHandle,uint,bool,bool) [{0x7fd3601edec0} + 0x163] [/Users/builder/jenkins/workspace/build-package-osx-mono/master@2/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Threading/WaitHandle.cs :: 90u] (0x10b5f3d40 0x10b5f3f43) [0x7fd35ef0e520 - VisualStudio.exe]
!   :             11076 (wrapper managed-to-native) System.Threading.WaitHandle:Wait_internal (intptr*,int,bool,int) [{0x7fd3601eeaf0} + 0x9b]  (0x10b5f4070 0x10b5f418d) [0x7fd35ef0e520 - VisualStudio.exe]
!   :               11076 ves_icall_System_Threading_WaitHandle_Wait_internal_raw  (in libmonosgen-2.0.dylib) + 88  [0x105bc1dc8]  icall-def.h:1148
```

**Comment by Therzok** [comment]
If we don't have that, we'll end up using way more CPU time due to no caching. We might need to make it async.

**Comment by Therzok** [comment]
We should probably not be calling the `GetDirectoryVersionInfo` here with Recursive=true, but rather do one single-depth `GetDirectoryVersionInfo` for each path in `files`

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-21] Comment by **xamarinc**



Fixed in version 8.0.0.2792 (release-8.0)



Pull Request #323 merged by: Vsevolod Kukol  
 Author: xamarin  
 Commit: 3fb561f057d82cec6dd3f302180218955b49d774 (xamarin/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-23] Comment by **xamarinc**



Fixed in version 8.1.0.1701 (master-vnext)



Pull Request #323 merged by: Vsevolod Kukol  
 Author: xamarin  
 Commit: 3fb561f057d82cec6dd3f302180218955b49d774 (xamarin/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-32] Comment by **xamarinc**



Fixed in version 8.6.0.802 (master)



Pull Request #323 merged by: Vsevolod Kukol  
 Author: xamarin  
 Commit: 3fb561f057d82cec6dd3f302180218955b49d774 (xamarin/monodevelop)
